README.md
# Cloudsmith Maven Wagon

The Cloudsmith [Maven Wagon](http://maven.apache.org/wagon/) provides an
integration between Cloudsmith and [Maven](https://maven.apache.org/),
[Gradle](https://gradle.org/), [SBT/Scala](https://www.scala-sbt.org/), and
more, for automated native/in-tool uploads of your artefacts. The configuration
for Maven, Gradle and SBT vary, so please refer to the relevant section for
details and examples.

The Cloudsmith Maven Wagon library isn't available on [Maven Central](https://search.maven.org/)
yet (but we're working on it). You can access the latest releases for the
library on the public [Cloudsmith API libraries repository](https://cloudsmith.io/package/ns/cloudsmith/repos/api/packages/).


## How Does It Work

The Maven Wagon utilises the [Cloudsmith API](https://api.cloudsmith.io) to push your packages to Cloudsmith.

The flow of this as follows:

1. Maven builds your artefacts (JAR, POM, etc.) using the standard build process.
2. Maven starts the deploy process by delegating to this library/plugin.
3. The library authenticates with the Cloudsmith API via HTTPS to check your credentials.
4. The library asks the Cloudsmith API via HTTPS for existing repository metadata.
5. The library checks to see if your artefacts are already located in the upstream repository.
6. Assuming your artefacts are new, the library performs the following for each artefacts:
  a. It requests upload authorisation from the Cloudsmith API via HTTPS.
  b. It receives the pre-signed upload request details for uploading the file.
  c. It uses these to push the artefact to S3 via HTTPS (the actual file upload).
7. When all artefacts uploaded, the library tells the Cloudsmith API via HTTPS that the uploads are completed.
8. The library uses the status endpoint of the Cloudsmith API to wait for the package to synchronise.
9. The library informs Maven of the success or failure of the deployment.

As noted above HTTPS is used to communicate with the Cloudsmith API and S3 - If you encounter issues with
timeouts, please refer to the HTTP Timeout Configuration below for help.


## Common

### Authentication Configuration

#### Cloudsmith API Key

In order for the library to communicate with Cloudsmith, you'll need to obtain your Cloudsmith API Key.

For convenience you can fetch this using the [Cloudsmith CLI](https://github.com/cloudsmith-io/cloudsmith-cli):

```shell
pip install cloudsmith-cli
cloudsmith token
```

You can also get it via the [User API Tokens on the Cloudsmith Website](https://cloudsmith.io/settings/api-tokens/) (requires sign-in).

*Note:* If you're automating upload via a CI/CD system, we recommend creating a least-privilege bot user for this task.


### Debug Output

If you need help with debugging issues you can enable debug output via (in order of precedence):

1. Set the `CLOUDSMITH_DEBUG` environment variable to `true`.
2. Set the `cloudsmith.debug` property to `true`.


### HTTP Timeouts

The library uses HTTPS to communicate with the Cloudsmith API and S3 to coordinate
package creation and upload of the package artefacts. This is usually problem-free,
but if you have an unstable or slow connection you may encounter timeouts during the
retrieval or upload portions of the deployment workflow. If this happens, you can
configure larger values for the HTTP timeouts, as below.

#### HTTP Timeout Configuration

The following defaults are used:

- HTTP connect timeout: 15 seconds
- HTTP read timeout: 30 seconds
- HTTP write timeout: 120 seconds

Configuring the HTTP connect timeout (in order of precedence):

1. Set the `CLOUDSMITH_HTTP_CONNECT_TIMEOUT` environment variable to a positive integer value (e.g. `30` for 30 seconds).
2. Set the `cloudsmith.http.connect.timeout` property to a positive integer value (e.g. `60` for 60 seconds).

Configuring the HTTP read timeout (in order of precedence):

1. Set the `CLOUDSMITH_HTTP_READ_TIMEOUT` environment variable to a positive integer value (e.g. `60` for 60 seconds).
2. Set the `cloudsmith.http.read.timeout` property to a positive integer value (e.g. `120` for 120 seconds).

Configuring the HTTP write timeout (in order of precedence):

1. Set the `CLOUDSMITH_HTTP_WRITE_TIMEOUT` environment variable to a positive integer value (e.g. `240` for 240 seconds).
2. Set the `cloudsmith.http.write.timeout` property to a positive integer value (e.g. `480` for 480 seconds).


### Synchronisation Wait

Packages that are uploaded to Cloudsmith are "eventually consistent" - By this
we mean that a package that is uploaded isn't instantaneous made available
(published) to users of the target repository. This is due to Cloudsmith using
background workers to process packages after they are uploaded, and after
a small back-off period these will begin to process packages as they are
uploaded.

For this reason, when a package is uploaded, by default the library will
initiate a mechanism to wait on the status of the uploaded package until
synchronisation is complete (or has failed).

The following is an example of what this looks like when package synchronisation is enabled:

```shell
[INFO] Creating a new Maven Package ...
[INFO] Waiting for the package to synchronise ...
[INFO] Status = Sync In Progress , Stage = Preparing for Synch , Progress = 15
[INFO] Status = Sync In Progress , Stage = Verifying Package File(s) , Progress = 35
[INFO] Status = Sync In Progress , Stage = Verifying Package File(s) , Progress = 35
[INFO] Status = Sync In Progress , Stage = Verifying Package File(s) , Progress = 35
[INFO] Status = Sync In Progress , Stage = Verifying Package File(s) , Progress = 35
[INFO] Status = Sync In Progress , Stage = Indexing Package File(s) , Progress = 65
[INFO] Status = Sync In Progress , Stage = Synching Repository , Progress = 85
[INFO] Status = Sync Completed , Stage = Fully Synchronised , Progress = 100
[INFO] Created: cloudsmith/examples/cloudsmith-maven-example-001-snapshotjar-2
```

#### Synchronisation Wait Configuration

By default synchronisation is enabled. If you'd prefer to not wait for package
synchronisation, with the caveat that you'll not be able to tell if the
synchronisation process succeeds or not, you can disable it in on the following
ways (in order of precedence):

1. Set the `CLOUDSMITH_SYNC_WAIT_ENABLED` environment variable to `false`.
2. Set the `cloudsmith.sync_wait.enabled` property to `false`.

If synchronisation wait is disabled, then the above example output will instead be something like:

```shell
[INFO] Creating a new Maven package ...
[INFO] Not waiting on the package to synchronise (note: failures won't be reported here)
[INFO] Created: cloudsmith/examples/cloudsmith-maven-example-001-snapshotjar-3
```

#### Synchronisation Wait Verbosity Configuration

By default synchronisation is verbose. If you'd prefer to turn off details of
synchronisation, but still want to wait on the synchronisation to complete or
fail (along with a status message if it does), then you turn off verbosity in
the following ways (in order of precedence):

1. Set the `CLOUDSMITH_SYNC_WAIT_VERBOSE` environment variable to `false`.
2. Set the `cloudsmith.sync_wait.verbose` property to `false`.

#### Synchronisation Wait Interval Configuration

By default the synchronisation waits for an interval of 5000ms (5 seconds)
between updates. If you'd like to shorten or length this interval, then you can
set it in the following ways (in order of precedence):

1. Set the `CLOUDSMITH_SYNC_WAIT_INTERVAL` environment variable to a positive integer value (e.g. `20` for 20 seconds).
1. Set the `cloudsmith.sync_wait.interval` property to an positive integer value (e.g. `10` to 10 seconds).

*Note:* If you set it to zero or a negative number, then the synchronisation wait process will be disabled.


## Maven

### Deployment Configuration

#### Library Repository

The Cloudsmith Maven Wagon library isn't available on [Maven Central](https://search.maven.org/) yet (but we're working on it).

Until it is, you'll need to add the following configuration to your project `pom.xml`:

```xml
  <pluginRepositories>
    <pluginRepository>
      <id>cloudsmith-api</id>
      <name>Cloudsmith API Releases</name>
      <url>https://dl.cloudsmith.io/public/cloudsmith/api/maven</url>
    </pluginRepository>
  </pluginRepositories>

  <repositories>
    <repository>
      <id>cloudsmith-api</id>
      <name>Cloudsmith API Releases</name>
      <url>https://dl.cloudsmith.io/public/cloudsmith/api/maven</url>
    </repository>
  </repositories>
```

This will allow Maven to fetch it as a build/deploy dependency from Cloudsmith.

#### Library Dependency

To bring the library into your Maven project, add the following to your project `pom.xml` file within `<build>` and `<extensions>`:

```xml
  <build>
    <extensions>
      <extension>
        <groupId>io.cloudsmith.maven.wagon</groupId>
        <artifactId>cloudsmith-maven-wagon</artifactId>
        <version>0.4.0</version>
      </extension>
    </extensions>
  </build>
```

*Note:* Please refer to the [Cloudsmith API libraries repository](https://cloudsmith.io/package/ns/cloudsmith/repos/api/packages/) to replace *0.4.0* with the latest library version.

#### Upload Repositories

The upload repositories specify which Cloudsmith repository you'd like to upload your artefacts to.

To configure the upload repositories for your project, add the following to your project `pom.xml` file within `<distributionManagement>`:

```xml
  <distributionManagement>
    <snapshotRepository>
      <id>cloudsmith-snapshots</id>
      <url>cloudsmith+https://api.cloudsmith.io/your-namespace/your-snapshots-repo</url>
    </snapshotRepository>
    <repository>
      <id>cloudsmith-releases</id>
      <url>cloudsmith+https://api.cloudsmith.io/your-namespace/your-releases-repo</url>
    </repository>
  </distributionManagement>
```

Replacing the following terms with your own configuration:

- `your-namespace`: Replace with your user or organization slug.
- `your-snapshots-repo`: Replace with your snapshots (edge releases) repository slug.
- `your-releases-repo`: Replace with your releases (non-edge releases) repository slug.

*Note 1:* The repositories *must* exist prior to deployment - Create them first!

*Note 2:* You can configure the snapshots and releases repositories to be the same, they do not need to be different.

*Note 3:* You can replace `cloudsmith-snapshots` and `cloudsmith-releases` with your own identifiers.

### Authentication Configuration

#### Cloudsmith API Key

Please see the common setup above to obtain your API Key for Cloudsmith.

#### Precedence

You can configure the library with your API Key in one of following ways (in order of precedence):

1. System Property
2. Environment Variable
3. User-specific Settings

#### System Property

You can set the `cloudsmith.api_key` system property with your API Key:

```xml
<properties>
  <cloudsmith.api_key>your-API-key</cloudsmith.api_key>
</properties>
```

Replacing the following terms with your own configuration:

- `your-API-key`: Your Cloudsmith user API key (see above for how to retrieve it).

**Note**: This applies to *all* repositories. If you need more granularity, use the user settings file approach below.

#### Environment Variable

You can export your API Key using the `CLOUDSMITH_API_KEY` environment variable, such as (Linux example):

```shell
export CLOUDSMITH_API_KEY=your-API-key

```

Replacing the following terms with your own configuration:

- `your-API-key`: Your Cloudsmith user API key (see above for how to retrieve it).

**Note**: This applies to *all* repositories. If you need more granularity, use the user settings file approach below.

#### User-specific Settings

You can configure your `$HOME/.m2/settings.xml` file with your Cloudsmith API Key:

```xml
<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
                      https://maven.apache.org/xsd/settings-1.0.0.xsd">
  <servers>
    <server>
      <id>cloudsmith-snapshots</id>
      <password>your-API-key</password>
    </server>
    <server>
      <id>cloudsmith-releases</id>
      <password>your-API-key</password>
    </server>
  </servers>
</settings>
```

Replacing the following terms with your own configuration:

- `your-API-key`: Your Cloudsmith user API key (see above for how to retrieve it).

*Note 1:* Rather than putting your API keys as plaintext, you can encrypt them using [Maven Password Encryption](https://maven.apache.org/guides/mini/guide-encryption.html).

*Note 2:* The `<id>` for each server needs to match those in your `pom.xml` file under `<distributionManagement>`.

*Note 3:* You can replace `cloudsmith-snapshots` and `cloudsmith-releases` with your own identifiers.

### Usage

Assuming you have authentication and configuration setup, as above, you'll be able to publish to Cloudsmith via:

```shell
mvn deploy
```

### Example Project

We have a fully-worked [example project for Maven](https://github.com/cloudsmith-io/cloudsmith-examples/tree/master/projects/maven/src) that you can use as a reference.

The output of this is uploaded to the publicly available [Cloudsmith examples repository](https://cloudsmith.io/package/ns/cloudsmith/repos/examples/packages/) as part of our testing processes.


## Gradle

### Note

The configuration for Gradle use the "old-style" Maven plugin documented in the [Gradle Maven Plugin documentation](https://docs.gradle.org/current/userguide/maven_plugin.html).

### Deployment Configuration

#### Library Repository

The Cloudsmith Maven Wagon library isn't available on [Maven Central](https://search.maven.org/) yet (but we're working on it).

Until it is, you'll need to add the following configuration to your project `build.gradle` file within `repositories`:

```groovy
repositories {
  maven {
    name = "Cloudsmith API Releases"
    url = "https://dl.cloudsmith.io/public/cloudsmith/api/maven"
  }

  // You might have references to mavenLocal() and mavenCentral() here too
}
```

This will allow Gradle to fetch it as a build/deploy dependency from Cloudsmith.

#### Library Dependency

To bring the library into your Gradle project, add the following to your project `build.gradle` file:

```groovy
apply plugin: 'maven'

configurations {
  deployerJars
}

dependencies {
  deployerJars 'io.cloudsmith.maven.wagon:cloudsmith-maven-wagon:0.4.0'
}
```

*Note:* Please refer to the [Cloudsmith API libraries repository](https://cloudsmith.io/package/ns/cloudsmith/repos/api/packages/) to replace *0.4.0* with the latest library version.


#### Upload Repositories

The upload repositories specify which Cloudsmith repository you'd like to upload your artefacts to.

To configure the upload repositories for your project, add the following to your project `build.gradle` file:

```groovy
uploadArchives {
   repositories {
     mavenDeployer {
       configuration = configurations.deployerJars

       repository(url: "cloudsmith+https://api.cloudsmith.io/your-namespace/your-releases-repo") {
         authentication(password: "$cloudsmithApiKey")
       }

       snapshotRepository(url: "cloudsmith+https://api.cloudsmith.io/your-namespace/your-snapshots-repo") {
         authentication(password: "$cloudsmithApiKey")
       }
   }
}
```

Replacing the following terms with your own configuration:

- `your-namespace`: Replace with your user or organization slug.
- `your-snapshots-repo`: Replace with your snapshots (edge releases) repository slug.
- `your-releases-repo`: Replace with your releases (non-edge releases) repository slug.

*Note 1:* The repositories *must* exist prior to deployment - Create them first!

*Note 2:* You can configure the snapshots and releases repositories to be the same, they do not need to be different.

*Note 3:* You need to define `$cloudsmithApiKey` as a gradle property. See below.

### Authentication Configuration

#### Cloudsmith API Key

Please see the common setup above to obtain your API Key for Cloudsmith.

#### Precedence

You can configure the library with your API Key in one of the following ways (in order of precedence):

1. Environment Variable
2. Gradle Property

#### Environment Variable

You can export your API Key using the `CLOUDSMITH_API_KEY` environment variable, such as (Linux example):

```
export CLOUDSMITH_API_KEY=your-API-key

```

Replacing the following terms with your own configuration:

- `your-API-key`: Your Cloudsmith user API key (see above for how to retrieve it).

**Note**: This applies to *all* repositories. If you need more granularity, use the user Gradle Property file approach below.

#### Gradle Property

You can configure your `gradle.properties` file with your Cloudsmith API Key:

```shell
cloudsmithApiKey=your-API-key
```

Replacing the following terms with your own configuration:

- `your-API-key`: Your Cloudsmith user API key (see above for how to retrieve it).

### Usage

Assuming you have authentication and configuration setup, as above, you'll be able to publish to Cloudsmith via:

```shell
gradle uploadArchives
```

### Example Project

We have a fully-worked [example project for Gradle](https://github.com/cloudsmith-io/cloudsmith-examples/tree/master/projects/gradle/src) that you can use as a reference.

The output of this is uploaded to the publicly available [Cloudsmith examples repository](https://cloudsmith.io/package/ns/cloudsmith/repos/examples/packages/) as part of our testing processes.


## Scala/SBT

### Deployment Configuration

#### Library Dependency

Cloudsmith uses [sbt-aether-deploy](https://github.com/arktekk/sbt-aether-deploy) to wrap/use this library for publishing within SBT.

To bring the library into your SBT project, add the following to your project `project/plugins.sbt` file:

```
addSbtPlugin("no.arktekk.sbt" % "aether-deploy" % "0.20.0")
libraryDependencies += "io.cloudsmith.maven.wagon" % "cloudsmith-maven-wagon" % "0.4.0"
resolvers += Resolver.mavenLocal
```

*Note1 :* They have removed the WagonWrapper functionality (we think temporarily) from 0.21.0 of aether-deploy so use 0.20.0 - it is supported and available from maven central and our cloudsmith/api public repository.

*Note 2:* Please refer to the [Cloudsmith API libraries repository](https://cloudsmith.io/package/ns/cloudsmith/repos/api/packages/) to replace *0.4.0* with the latest library version.


#### Upload Repositories

The upload repositories specify which Cloudsmith repository you'd like to upload your artefacts to.

To configure the upload repositories for your project, add the following to your project `build.sbt` file:

```
credentials += Credentials(Path.userHome / ".sbt" / ".credentials")
aetherWagons := Seq(aether.WagonWrapper("cloudsmith+https", "io.cloudsmith.maven.wagon.CloudsmithWagon"))
publishTo := {
    Some("cloudsmith+https" at "cloudsmith+https://api.cloudsmith.io/your-namespace/your-repo")
}
```
Replacing the following terms with your own configuration:
- `your-namespace`: Replace with your user or organization slug.
- `your-repo`: Replace with your repository slug.

*Note:* The repositories must exist prior to deployment - Create them first!

### Authentication Configuration

#### Cloudsmith API Key

Please see the common setup above to obtain your API Key for Cloudsmith.

#### Sbt Property

Configure your `.credentials` file with the following details:

```
realm = cloudsmith
host = api.cloudsmith.io
user =
password = your-API-key
```

Replacing the following terms with your own configuration:

- `your-API-key`: Your Cloudsmith user API key (see above for how to retrieve it).

### Usage
Assuming you have authentication and configuration setup, as above, you'll be able to publish to Cloudsmith via:

```
sbt aetherDeploy
```

### Example Project

We have a fully-worked [example project for SBT](https://github.com/cloudsmith-io/cloudsmith-examples/tree/master/projects/sbt/src) that you can use as a reference.

The output of this is uploaded to the publicly available [Cloudsmith examples repository](https://cloudsmith.io/package/ns/cloudsmith/repos/examples/packages/) as part of our testing processes.


## Leiningen

Details coming soon.


## Kotlin/Kobalt

Details coming soon.


## License

Copyright 2018 Cloudsmith Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


## EOF

This quality product was brought to you by [Cloudsmith](https://cloudsmith.io) and the [fine folks who have contributed](https://github.com/cloudsmith-io/cloudsmith-maven-wagon/blob/master/CONTRIBUTORS.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

**Note:** Until 1.0 any MAJOR or MINOR release may have backwards-incompatible changes.

## [Unreleased]

Nothing yet.


## [0.4.0] - 2018-08-14

### Added

- Add support for multi-module/sub-project POMs.


## [0.3.1] - 2018-03-05

### Fixed

- Issue with sync wait treating seconds as millis (making wait non-existant).


## [0.3.0] - 2018-02-28

### Added

- Configuration for HTTP connect, read and write timeouts.
- Documentation for enabling debug via `CLOUDSMITH_DEBUG` (env) or `cloudsmith.debug` (property).
- Documentation for how the Maven Wagon actually works and interacts with the Cloudsmith API.

### Changed

- Prefix for synchronisation wait properties changed to `cloudsmith.sync_wait` prefix (not backwards compatible).
- Property for synchronisation wait verbose has dropped the `_ENABLED` suffix.


## [0.2.0] - 2018-02-02

### Added

- More verbosity for synchronisation wait.
- Support for disabling synchronisation wait.
- Support for changing synchronisation wait interval.

### Fixed

- Support for Java archives other than JAR, so EAR/WAR/AAR/etc. will now work.


## [0.1.0] - 2018-01-31

Phase 1 release (initial release).


scripts/test.sh
#!/usr/bin/env bash
self=$(readlink -f $BASH_SOURCE)
self_dir=$(dirname $self)
root_dir=$(readlink -f "$self_dir/..")
language=$1
api_url=$2
. $root_dir/scripts/common.sh

mvn test


src/main/java/io/cloudsmith/maven/wagon/CloudsmithRepository.java
package io.cloudsmith.maven.wagon;

import java.net.URL;
import java.net.MalformedURLException;
import java.util.Arrays;

import org.apache.maven.wagon.repository.Repository;


/**
 * Holds additional information about upstream Cloudsmith repos.
 */
@SuppressWarnings("serial")
public class CloudsmithRepository {
    private static final String PREFIX = "cloudsmith";
    private static final String DEFAULT_HOST = "api.cloudsmith.io";
    private static final String DEFAULT_PATH = "";

    private Repository repository = null;
    private String apiUrl;
    private String ownerName;
    private String repositoryName;

    public CloudsmithRepository(Repository repository) {
        this.repository = repository;
        determineCloudsmithInfo();
    }

    public String getOwnerName() {
        return this.ownerName;
    }

    public String getRepositoryName() {
        return this.repositoryName;
    }

    public String getApiUrl() {
        return this.apiUrl;
    }

    // Helpers

    private void determineCloudsmithInfo() {
        /* Form: cloudsmith://<api_url>/namespace/repo/ */
        String baseUrl = repository.getUrl();

        String[] baseParts = baseUrl.split("\\+");
        if (baseParts.length != 2 || !baseParts[0].equals(PREFIX)) {
            raiseCloudsmithInfoException(baseUrl);
        }

        URL url = null;

        try {
            url = new URL(baseParts[1]);
        } catch (MalformedURLException ex) {
            raiseCloudsmithInfoException(baseUrl);
        }

        String host = DEFAULT_HOST;

        if (url.getHost() != null && !url.getHost().equals(PREFIX)) {
            host = url.getHost();
        }

        String path = DEFAULT_PATH;
        String[] pathParts = url.getPath().split("/");
        if (pathParts.length < 2) {
            raiseCloudsmithInfoException(baseUrl);
        }

        if (pathParts.length > 2) {
            path = String.join(
                "/", Arrays.copyOfRange(pathParts, 0, pathParts.length - 2)
            );
        }

        try {
            url = new URL(url.getProtocol(), host, url.getPort(), path);
        } catch (MalformedURLException ex) {
            raiseCloudsmithInfoException(baseUrl);
        }

        this.apiUrl = url.toString();
        this.ownerName = pathParts[pathParts.length - 2];
        this.repositoryName = pathParts[pathParts.length - 1];
    }

    private void raiseCloudsmithInfoException(String baseUrl) {
        throw new IllegalArgumentException(
           "Repository <url> must be in form of "
            + "'cloudsmith+https://<api_url>/<namespace>/<repo>', got: "
            + baseUrl
        );
    }
}


src/main/java-templates/io/cloudsmith/maven/wagon/Properties.java

package io.cloudsmith.maven.wagon;

public final class Properties {
    //
    // Static Properties
    //

    private static final String  API_VERSION  = "${cloudsmith.api_version}";
    private static final String  ARTIFACT_ID  = "${project.artifactId}";
    private static final String  BUILD_NUMBER = "${buildNumber}";
    private static final String  GROUP_ID     = "${project.groupId}";
    private static final String  VERSION      = "${project.version}";
    private static final String  URL          = "${project.url}";

    //
    // Runtime Properties
    //

    private static final String  API_KEY_ENVVAR                = "CLOUDSMITH_API_KEY";
    private static final String  API_KEY_PROPERTY              = "cloudsmith.api_key";
    private static final boolean DEBUG_DEFAULT                 = false;
    private static final String  DEBUG_ENVVAR                  = "CLOUDSMITH_DEBUG";
    private static final String  DEBUG_PROPERTY                = "cloudsmith.debug";
    private static final int     HTTP_CONNECT_TIMEOUT_DEFAULT  = 15; // seconds
    private static final String  HTTP_CONNECT_TIMEOUT_ENVVAR   = "CLOUDSMITH_HTTP_CONNECT_TIMEOUT";
    private static final String  HTTP_CONNECT_TIMEOUT_PROPERTY = "cloudsmith.http.connect.timeout";
    private static final int     HTTP_READ_TIMEOUT_DEFAULT     = 30; // seconds
    private static final String  HTTP_READ_TIMEOUT_ENVVAR      = "CLOUDSMITH_HTTP_READ_TIMEOUT";
    private static final String  HTTP_READ_TIMEOUT_PROPERTY    = "cloudsmith.http.get.timeout";
    private static final int     HTTP_WRITE_TIMEOUT_DEFAULT    = 120; // seconds
    private static final String  HTTP_WRITE_TIMEOUT_ENVVAR     = "CLOUDSMITH_HTTP_WRITE_TIMEOUT";
    private static final String  HTTP_WRITE_TIMEOUT_PROPERTY   = "cloudsmith.http.put.timeout";
    private static final boolean SW_ENABLED_DEFAULT            = true;
    private static final String  SW_ENABLED_ENVVAR             = "CLOUDSMITH_SYNC_WAIT_ENABLED";
    private static final String  SW_ENABLED_PROPERTY           = "cloudsmith.sync_wait.enabled";
    private static final int     SW_INTERVAL_DEFAULT           = 5; // seconds
    private static final String  SW_INTERVAL_ENVVAR            = "CLOUDSMITH_SYNC_WAIT_INTERVAL";
    private static final String  SW_INTERVAL_PROPERTY          = "cloudsmith.sync_wait.interval";
    private static final boolean SW_VERBOSE_DEFAULT            = true;
    private static final String  SW_VERBOSE_ENVVAR             = "CLOUDSMITH_SYNC_WAIT_VERBOSE";
    private static final String  SW_VERBOSE_PROPERTY           = "cloudsmith.sync_wait.verbose";

    //
    // Static Properties
    //

    public static String getApiVersion() {
        return API_VERSION;
    }

    public static String getArtifactId() {
        return ARTIFACT_ID;
    }

    public static String getBuildNumber() {
        return BUILD_NUMBER;
    }

    public static String getGroupId() {
        return GROUP_ID;
    }

    public static String getVersion() {
        return VERSION;
    }

    public static String getUrl() {
        return URL;
    }

    //
    // Runtime Properties
    //

    public static String getApiKey(String defaultValue) {
        return getStringValue(
            API_KEY_ENVVAR,
            API_KEY_PROPERTY,
            defaultValue
        );
    }

    public static boolean isDebug() {
        return getBooleanValue(
            DEBUG_ENVVAR,
            DEBUG_PROPERTY,
            DEBUG_DEFAULT
        );
    }

    public static int getHttpConnectTimeout() {
        int value = getIntegerValue(
            HTTP_CONNECT_TIMEOUT_ENVVAR,
            HTTP_CONNECT_TIMEOUT_PROPERTY,
            HTTP_CONNECT_TIMEOUT_DEFAULT
        );

        if (value < 1) {
            System.out.println(
                "[WARN] HTTP connect timeout cannot be less than 1, setting value to 1."
            );
            value = 1;
        }

        return value;
    }

    public static int getHttpReadTimeout() {
        int value = getIntegerValue(
            HTTP_READ_TIMEOUT_ENVVAR,
            HTTP_READ_TIMEOUT_PROPERTY,
            HTTP_READ_TIMEOUT_DEFAULT
        );

        if (value < 1) {
            System.out.println(
                "[WARN] HTTP read timeout cannot be less than 1, setting value to 1."
            );
            value = 1;
        }

        return value;
    }

    public static int getHttpWriteTimeout() {
        int value = getIntegerValue(
            HTTP_WRITE_TIMEOUT_ENVVAR,
            HTTP_WRITE_TIMEOUT_PROPERTY,
            HTTP_WRITE_TIMEOUT_DEFAULT
        );

        if (value < 1) {
            System.out.println(
                "[WARN] HTTP write timeout cannot be less than 1, setting value to 1."
            );
            value = 1;
        }

        return value;
    }

    public static boolean isSyncWaitEnabled() {
        return getBooleanValue(
            SW_ENABLED_ENVVAR,
            SW_ENABLED_PROPERTY,
            SW_ENABLED_DEFAULT
        );
    }

    public static int getSyncWaitInterval() {
        int value = getIntegerValue(
            SW_INTERVAL_ENVVAR,
            SW_INTERVAL_PROPERTY,
            SW_INTERVAL_DEFAULT
        );

        if (value < 1) {
            System.out.println(
                "[WARN] Sync wait interval cannot be less than 1, setting value to 1."
            );
            value = 1;
        }

        return value;
    }

    public static boolean isSyncWaitVerbose() {
        return getBooleanValue(
            SW_VERBOSE_ENVVAR,
            SW_VERBOSE_PROPERTY,
            SW_VERBOSE_DEFAULT
        );
    }

    //
    // Private Helpers
    //

    private static String getStringValue(String envvar, String property, String defaultValue) {
        String value;

        if (envvar != null && !envvar.isEmpty()) {
            value = System.getenv(envvar);
            if (value != null && !value.isEmpty()) {
                return value;
            }
        }

        if (property != null && !property.isEmpty()) {
            value = System.getProperty(property);
            if (value != null && !value.isEmpty()) {
                return value;
            }
        }

        return defaultValue;
    }

    public static boolean getBooleanValue(String envvar, String property, boolean defaultValue) {
        String value = getStringValue(envvar, property, null);
        if (value == null || value.isEmpty()) {
            return defaultValue;
        }

        return Boolean.parseBoolean(value);
    }

    public static int getIntegerValue(String envvar, String property, int defaultValue) {
        String value = getStringValue(envvar, property, null);
        if (value == null || value.isEmpty()) {
            return defaultValue;
        }

        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException ex) {
            return defaultValue;
        }
    }

    public static double getDoubleValue(String envvar, String property, double defaultValue) {
        String value = getStringValue(envvar, property, null);
        if (value == null || value.isEmpty()) {
            return defaultValue;
        }

        try {
            return Double.parseDouble(value);
        } catch (NumberFormatException ex) {
            return defaultValue;
        }
    }

    private Properties() {
        throw new AssertionError("Instantiating utility class.");
    }
}
